File: prototyping/landscape.rb
require_relative "bot.rb"

class Landscape
  attr_reader :height, :width, :bots
  attr_accessor :grid

  def initialize(width:, height:, density_percent: 20)
    @width = width
    @height = height
    @density_percent = density_percent
    initialize_grid
  end

  def initialize_grid
    @bots = []
    @grid = Array.new(@height) { Array.new(@width) }
    coords = []
    markers = ("A".."Z").to_a
    @height.times { |r| @width.times { |c| coords << {r: r, c: c} } }
    num_of_bots = (coords.length * @density_percent / 100.0).to_i
    coords.shuffle.first(num_of_bots).each do |coord|
      add_bot(c: coord[:c], r: coord[:r], marker: markers.sample, algorithm: "right")
    end
  end

  def add_bot(params)
    bot = Bot.new(params)
    bot.landscape = self
    @bots << bot
    @grid[bot.r][bot.c] = bot
  end

  def update_position_for(bot)
    4.times do
      if bot.is_about_to_collide?
        bot.change_direction!
      else
        @grid[bot.r][bot.c] = nil
        bot.move_forward!
        @grid[bot.r][bot.c] = bot
        break
      end
    end
  end

  def prepare_frame
    @bots.each { |bot| update_position_for(bot) }
  end

  def render_frame
    system('clear')
    @grid.each do |row|
      print "|"
      row.each { |bot| print (bot ? "#{bot.marker}|" : " |") }
      puts
    end
  end
end

File: js/landscape.js
var stampit = require('stampit')
var Fixed2DArray = require('fixed-2d-array')
var _ = require('lodash')
var $ = require('jquery')
var Bot = require('./bot')
var randomInt = require('random-int')

var Landscape = stampit({
  init: function () {
    this.$element = $('#landscape')
    this.$botCounter = $('#bot-count')
    this.$collisionAvoidedCounter = $('#collisions-avoided-count')
    this.collisionsAvoided = 0
    this.grid = new Fixed2DArray(this.height, this.width, null)
    this.bots = []
    this.initializeBots()
    this.render()
  },
  methods: {
    addBot: function () {
      var bot = Bot(_.merge(this.getRandCoords(), {scale: this.scale}))
      bot.landscape = this
      this.bots.push(bot)
      this.grid.set(bot.r, bot.c, bot)
      bot.render()
      this.$botCounter.text(this.bots.length)
    },
    updateFrame: function () {
      _.each(this.bots, this.updatePositionFor.bind(this))
    },

    // private
    initializeBots: function () {
      var numOfCells = this.width * this.height
      var numOfBots = parseInt(numOfCells * this.densityPercent / 100)
      _.times(numOfBots, this.addBot.bind(this))
    },
    getRandCoords: function () {
      return { r: randomInt(this.height - 1), c: randomInt(this.width - 1) }
    },
    render: function () {
      this.$element.css({
        width: this.width * this.scale,
        height: this.height * this.scale,
      })
    },
    updatePositionFor: function (bot) {
      for (var i = 0; i < 4; i++) {
        if (bot.isAboutToCollide()) {
          bot.changeDirection()
          if (bot.isAlive()) { bot.dieSlowly() }
          this.collisionsAvoided++
          this.$collisionAvoidedCounter.text(this.collisionsAvoided)
        } else {
          this.grid.set(bot.r, bot.c, null)
          bot.moveForward()
          this.grid.set(bot.r, bot.c, bot)
          break
        }
      }
    }
  }
})

module.exports = Landscape
